site/app/controllers/pdf/PDFController.php

```diff
@@ -113,13 +101,20 @@ public function savePDFAnnotation($gradeable_id) {
         if (!is_dir($annotation_version_path) && !FileUtils::createDir($annotation_version_path)) {
             return $this->core->getOutput()->renderJsonFail('Creating annotation version folder failed.');
         }
-
-        if (strpos($annotation_info['file_name'], '..') !== false) {
-            return false;
-        }
-
-        $new_file_name = preg_replace('/\\.[^.\\s]{3,4}$/', '', $annotation_info['file_name']) . "_" . $grader_id . '.json';
-        file_put_contents(FileUtils::joinPaths($annotation_version_path, $new_file_name), $annotation_layer);
+        
+        $partial_path = substr_replace($annotation_info["file_path"], "", 0, strlen($annotation_version_path) + 1);
+        
+        $annotation_layer_decoded = json_decode($_POST['annotation_layer'], true);
```

Comment:
what does this JSON object look like with two annotations?

Reply:
Only the first annotation is labeled with the grader_id and partial_path. I've updated to label each annotation with the partial path and grader_id.

Reply:
can you post a sample JSON object here of two annotations?

Reply:
Each of the annotations are labeled with the grader_id and file_path

Reply:
Github can't process the file for some reason, here are the contents [{"type":"drawing","width":5.5,"color":"#FF0000","lines":[["100.14","61.15"],["104.14","63.44"],["109.86","65.15"],["115.57","66.87"],["124.14","69.72"],["131.57","72.58"],["144.14","78.29"],["160.71","86.87"],["179.00","97.15"],["193.29","104.01"],["207.57","112.01"],["220.71","119.44"],["233.86","128.01"],["243.00","133.72"],["250.43","138.29"],["255.57","141.72"],["260.14","146.29"],["264.14","149.72"],["267.57","152.01"],["269.86","152.58"],["273.29","154.29"],["276.14","154.87"],["279.00","155.44"],["282.43","155.44"],["286.43","154.29"],["292.14","151.44"],["297.29","147.44"],["304.71","139.44"],["312.14","127.44"],["318.43","112.58"],["320.71","98.87"],["322.43","91.44"],["322.43","86.87"],["322.43","84.58"],["319.57","82.87"],["312.14","80.01"],["300.71","78.87"],["284.14","80.58"],["260.14","87.44"],["225.86","95.44"],["164.14","108.01"],["144.71","113.15"],["135.00","116.01"],["132.14","116.58"]],"class":"Annotation","uuid":"241c4a9b-646e-43d6-9a26-32255eeace4e","page":1,"userId":"instructor","file_path":"summary 21 1 page.pdf","grader_id":"instructor"},{"type":"drawing","width":5.5,"color":"#FF0000","lines":[["327.90","47.89"],["329.44","48.66"],["331.75","49.43"],["334.83","50.97"],["337.14","51.74"],["340.22","53.28"],["343.30","55.59"],["346.38","57.13"],["351.77","59.45"],["356.39","63.30"],["361.78","67.15"],["367.17","69.46"],["372.56","73.31"],["379.49","77.16"],["384.88","81.01"],["391.81","84.86"],["398.74","88.71"],["404.13","93.33"],["411.06","96.41"],["416.46","101.03"],["423.39","104.11"],["428.78","107.96"],["432.63","110.27"],["440.33","114.12"],["444.18","116.43"],["449.57","119.51"],["454.19","121.05"],["457.27","122.59"],["459.58","123.36"],["463.43","125.67"],["467.28","126.44"],["469.59","126.44"],["471.13","127.21"],["473.44","127.98"],["474.21","128.75"],["474.98","128.75"],["475.75","128.75"],["477.29","128.75"],["478.83","128.75"],["481.14","128.75"],["482.68","127.98"],["484.99","125.67"],["488.84","124.13"],["491.15","121.82"],["494.23","119.51"],["497.31","116.43"],["498.85","114.12"],["501.93","110.27"],["504.24","106.42"],["505.78","103.34"],["506.55","99.49"],["507.32","97.18"],["508.09","94.10"],["508.09","91.79"],["508.09","88.71"],["505.78","85.63"],["503.47","82.55"],["500.39","77.93"],["496.54","73.31"],["491.92","70.23"],["488.07","66.38"],["483.45","62.53"],["478.06","59.45"],["473.44","57.13"],["469.59","55.59"],["466.51","54.82"],["464.20","54.82"],["461.89","54.82"],["460.35","54.82"],["458.81","54.82"],["457.27","56.36"],["456.50","58.68"],["454.19","63.30"],["452.65","68.69"],["451.11","71.77"],["448.80","80.24"],["445.72","87.17"],["442.64","92.56"],["440.33","97.95"],["438.02","103.34"],["435.71","107.96"],["434.17","113.35"],["431.86","116.43"],["429.55","121.82"],["428.78","125.67"],["428.01","132.60"],["428.01","139.53"],["428.01","145.69"],["428.01","151.85"],["428.01","158.79"],["428.01","162.64"],["428.78","167.26"],["428.78","169.57"],["429.55","171.88"],["429.55","173.42"],["430.32","174.19"],["431.09","174.19"],["431.86","174.19"],["433.40","174.19"],["435.71","174.19"],["438.79","173.42"],["444.18","172.65"],["448.03","170.34"],["454.96","168.80"],["461.12","165.72"],["468.05","161.87"],["474.98","158.79"],["481.14","154.93"],["486.53","153.39"],["495.00","149.54"],["500.39","146.46"],["505.78","144.15"],["510.40","141.07"],["515.03","138.76"],["518.11","136.45"],["520.42","134.91"],["521.96","133.37"],["523.50","131.83"],["524.27","130.29"],["524.27","128.75"],["525.04","127.98"],["525.04","124.90"],["525.04","122.59"],["524.27","119.51"],["521.96","116.43"],["518.88","111.04"],["516.57","105.65"],["511.94","100.26"],["508.09","94.87"],["501.16","85.63"],["495.00","78.70"],["488.07","71.77"],["481.14","64.84"],["474.21","57.91"],["467.28","52.51"],["461.89","47.89"],["456.50","45.58"],["450.34","41.73"],["446.49","40.19"],["441.87","39.42"],["438.79","38.65"],["435.71","37.88"],["431.86","37.88"],["426.47","37.88"],["421.08","37.88"],["415.69","39.42"],["409.52","42.50"],["403.36","47.12"],["398.74","52.51"],["393.35","59.45"],["387.96","66.38"],["384.11","72.54"],["381.03","78.70"],["377.18","82.55"],["374.87","85.63"],["371.79","89.48"],["370.25","91.79"],["367.94","94.10"],["367.17","95.64"],["365.63","97.95"],["365.63","99.49"],["364.86","101.03"],["364.86","102.57"],["364.86","105.65"],["364.86","108.73"],["364.86","112.58"],["364.86","117.20"],["365.63","121.05"],["366.40","125.67"],["368.71","129.52"],["369.48","134.14"],["371.02","136.45"],["372.56","139.53"],["374.10","141.84"],["374.87","144.15"],["376.41","146.46"],["377.95","148.00"],["378.72","148.77"],["379.49","149.54"],["380.26","151.08"],["381.03","152.62"],["382.57","154.16"],["384.11","155.70"],["386.42","158.02"],["388.73","159.56"],["391.04","161.10"],["391.81","162.64"],["394.12","164.18"],["394.89","164.95"],["396.43","164.95"],["396.43","165.72"],["397.97","165.72"],["398.74","165.72"],["399.51","165.72"],["400.28","166.49"],["401.82","166.49"],["402.59","166.49"],["403.36","166.49"]],"class":"Annotation","uuid":"738f1083-2f89-4bbd-b4c1-00f2d715ef5e","page":1,"userId":"instructor","file_path":"summary 21 1 page.pdf","grader_id":"instructor"},{"type":"drawing","width":5.5,"color":"#FF0000","lines":[["200.06","25.56"],["201.60","25.56"],["203.14","26.33"],["205.45","26.33"],["208.53","27.10"],["211.61","27.87"],["216.24","28.64"],["221.63","29.41"],["225.48","29.41"],["231.64","31.72"],["237.80","32.49"],["244.73","34.03"],["252.43","34.80"],["260.90","36.34"],["268.60","37.88"],["277.07","40.19"],["284.77","40.96"],["292.47","42.50"],["300.17","44.04"],["307.87","44.81"],["314.04","45.58"],["319.43","46.35"],["324.82","48.66"],["329.44","48.66"],["331.75","49.43"],["334.06","50.20"],["336.37","50.20"],["337.14","50.97"]],"class":"Annotation","uuid":"7a2d1ea2-58a6-4f03-9868-31353a1ddb05","page":1,"userId":"instructor","file_path":"summary 21 1 page.pdf","grader_id":"instructor"}]

Reply:
Can we restructure this to be:
```
{
  "grader_id": "instructor",
  "file_path": "summary 21 1 page.pdf",
  "annotations: [ ... ]
}
```
We should not be modifying the annotation objects so that it's a lossless round-tripping between Submitty and pdf-annotate.js.

Reply:
I assume you meant to reply here with that error, and not in the thread about the regex.

Yes, you would need to also edit where we're loading the file into localStorage: https://github.com/Submitty/Submitty/blob/e004c21305c42251ebc147c128843ff130096a1b/site/app/templates/grading/electronic/PDFAnnotationEmbedded.twig#L20-L25

I would recommend looking over the [`LocalUserStoreAdapter`](https://github.com/Submitty/pdf-annotate.js/blob/master/src/adapter/LocalUserStoreAdapter.js) of the pdf-annotate.js library if you have not yet already.